RecoMuon/TrackerSeedGenerator/plugins/TSGForOIFromL2.cc

```diff
@@ -0,0 +1,474 @@
+
+  // Error Rescaling
+  TrajectoryStateOnSurface onLayer(tsos);
+  onLayer.rescaleError(errorSF);
+
+  std::vector< GeometricSearchDet::DetWithState > dets;
+  layer.compatibleDetsV(onLayer, propagatorAlong, *estimator, dets);
+
+  // Find Measurements on each DetWithState
+  LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::makeSeedsFromHits: Find measurements on each detWithState  " << dets.size() << endl;
+  std::vector<TrajectoryMeasurement> meas;
+  for (std::vector<GeometricSearchDet::DetWithState>::iterator it=dets.begin(); it!=dets.end(); ++it) {
+    MeasurementDetWithData det = measurementTracker->idToDet(it->first->geographicalId());
+    if (det.isNull()) continue;
+    if (!it->second.isValid()) continue;	// Skip if TSOS is not valid
+
+    std::vector <TrajectoryMeasurement> mymeas = det.fastMeasurements(it->second, onLayer, propagatorAlong, *estimator);	// Second TSOS is not used
+    for (std::vector<TrajectoryMeasurement>::const_iterator it2 = mymeas.begin(), ed2 = mymeas.end(); it2 != ed2; ++it2) {
+      if (it2->recHit()->isValid()) meas.push_back(*it2);	// Only save those which are valid
+    }
+  }
+  
+  // Update TSOS using TMs after sorting, then create Trajectory Seed and put into vector
+  LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::makeSeedsFromHits: Update TSOS using TMs after sorting, then create Trajectory Seed, number of TM = " << meas.size() << endl;
+  std::sort(meas.begin(), meas.end(), TrajMeasLessEstim());
+
+  unsigned int found = 0;
+  for (std::vector<TrajectoryMeasurement>::const_iterator it=meas.begin(); it!=meas.end(); ++it) {
+    TrajectoryStateOnSurface updatedTSOS = updator_->update(it->forwardPredictedState(), *it->recHit());
+    LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::makeSeedsFromHits: TSOS for TM " << found << endl;
+    if (not updatedTSOS.isValid()) continue;
+
+    
+    edm::OwnVector<TrackingRecHit> seedHits;
+    seedHits.push_back(*it->recHit()->hit());
+    PTrajectoryStateOnDet const& pstate = trajectoryStateTransform::persistentState(updatedTSOS, it->recHit()->geographicalId().rawId());
+    TrajectorySeed seed(pstate, std::move(seedHits), oppositeToMomentum);
+    LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::makeSeedsFromHits: Number of seedHits: " << seedHits.size() << endl;
+    out.push_back(seed);
+    found++;
+    numSeedsMade++;
+    hitSeedsMade++;
+    if (found == numOfHitsToTry_) break;
+    if (hitSeedsMade > maxHitSeeds_) return;
+  }
+ 
+  if (found) layerCount++;
+
+}
+
+
+//
+// Calculate the dynamic error SF by analysing the L2
+//
+double TSGForOIFromL2::calculateSFFromL2(const reco::TrackRef track) const {
+    
+  double theSF = 1.0;
+  // L2 direction vs pT blowup - as was previously done:
+  // Split into 4 pT ranges: <pT1_, pT1_<pT2_, pT2_<pT3_, <pT4_: 13,30,70
+  // Split into different eta ranges depending in pT
+  double abseta = std::abs(track->eta());
+  if (track->pt() <= pT1_) theSF = SF1_;
+  if (track->pt() > pT1_ && track->pt() <= pT2_) {
```

Comment:
It looks to me that all requirements on pt and abseta are here mutually exclusive.
Therefore, ```else if``` could be used extensively in order to avoid useless checks